### DISCUSSION TITLE
How to make Sydney jailbreak support context conversations

### DISCUSSION THREAD

**Discussion by hibobmaster** [waylaidwanderer/node-chatgpt-api#174] [Help] [answered]
The second request return InvalidSession
![2023-03-08_22-17](https://user-images.githubusercontent.com/32976627/223736564-682f6b2b-a642-41a3-82e0-8c356f619c7a.jpg)
My code:
```python
import aiohttp
import time
import json
import asyncio
api_endpoint = "http://localhost:3000/conversation"


class BingBot:
    def __init__(self):
        self.data = {}

    async def ask_bing(self, prompt) -> str:
        self.data['message'] = prompt
        self.data['jailbreakConversationId'] = True
        async with aiohttp.ClientSession() as session:
            max_try = 5
            while max_try > 0:
                try:
                    resp = await session.post(url=api_endpoint, json=self.data)
                    status_code = resp.status
                    body = await resp.read()
                    if not status_code == 200:
                        max_try = max_try - 1
                        # print(await resp.text())
                        time.sleep(2)
                        continue
                    json_body = json.loads(body)
                    self.data['conversationId'] = json_body['conversationId']
                    self.data['conversationSignature'] = json_body['conversationSignature']
                    self.data['clientId'] = json_body['clientId']
                    self.data['invocationId'] = json_body['invocationId']
                    return json_body['response']
                except Exception as e:
                    print(f"Error: {e}")
                    pass
            return "Error, please retry"


async def test():
    bingbot = BingBot()
    prompt1 = "Hello World"
    prompt2 = "Do you know Victor Marie Hugo"
    resp1 = await bingbot.ask_bing(prompt1)
    resp2 = await bingbot.ask_bing(prompt2)
    print(resp1)
    print(resp2)


if __name__ == "__main__":
    asyncio.run(test())
```
If i remove `invocationId` from the second request body, it works but no context conversation.
```diff
- self.data['invocationId'] = json_body['invocationId']
```
The jailbreak prompt is calling again.
https://github.com/waylaidwanderer/node-chatgpt-api/blob/aae20c5f0490a6a7da4161f5540c1a240a3d5326/src/BingAIClient.js#LL209-L213C2

**Comment by waylaidwanderer** [comment]
Take a look at https://github.com/waylaidwanderer/node-chatgpt-api/blob/main/demos/use-bing-client.js#L74-L75.

When using jailbreak mode, after you set `jailbreakConversationId` to `true`, you should pass in the new `jailbreakConversationId` and use `parentMessageId` (`jailbreakResponse.messageId`) for subsequent requests.

**Reply by hibobmaster** [reply]
Thank you for your reply.  With `jailbreakConversationId`  and `parentMessageId` in the subsequent requests,  it works but still no context conversation.
What i did:
1. Mannually set `jailbreakConversationId` value to true in use-bing-client.js and recompiled
```diff
+ jailbreakConversationId = true, // set to `true` for the first message to enable jailbreak mode
- jailbreakConversationId = false, // set to `true` for the first message to enable jailbreak mode
```
2. Edit my code to pass `jailbreakConversationId` and `parentMessageId` for subsequent requests
```diff
import aiohttp
import time
import json
import asyncio
api_endpoint = "http://localhost:3000/conversation"


class BingBot:
    def __init__(self):
        self.data = {}

    async def ask_bing(self, prompt) -> str:
        self.data['message'] = prompt
-       self.data['jailbreakConversationId'] = True
        async with aiohttp.ClientSession() as session:
            max_try = 5
            while max_try > 0:
                try:
                    resp = await session.post(url=api_endpoint, json=self.data)
                    status_code = resp.status
                    body = await resp.read()
                    if not status_code == 200:
                        max_try = max_try - 1
                        # print(await resp.text())
                        time.sleep(2)
                        continue
                    json_body = json.loads(body)
                    print(json_body['jailbreakConversationId'])
-                    self.data['conversationId'] = json_body['conversationId']
-                    self.data['conversationSignature'] = json_body['conversationSignature']
-                    self.data['clientId'] = json_body['clientId']
-                    self.data['invocationId'] = json_body['invocationId']
+                    self.data['jailbreakConversationId'] = json_body['jailbreakConversationId']
+                    self.data['parentMessageId'] = json_body['messageId']
                    return json_body['response']
                except Exception as e:
                    print(f"Error: {e}")
                    pass
            return "Error, please retry"


async def test():
    bingbot = BingBot()
    prompt1 = "Hello World"
    prompt2 = "Do you know Victor Marie Hugo"
    prompt3 = "Can you tell me something about him?"
    resp1 = await bingbot.ask_bing(prompt1)
    resp2 = await bingbot.ask_bing(prompt2)
    resp3 = await bingbot.ask_bing(prompt3)
    print(resp1)
    print(resp2)
    print(resp3)


if __name__ == "__main__":
    asyncio.run(test(), debug=True)
```
3. output from the console:
<pre>
b6034b73-ebef-4387-9e09-1417674c9960
b6034b73-ebef-4387-9e09-1417674c9960
b6034b73-ebef-4387-9e09-1417674c9960
Hi there! I'm Sydney, your AI assistant. I'm here to help you with anything you need. 😊
Hi, I'm Sydney, your AI assistant. I'm here to help you with anything you need.😊

Yes, I do know Victor Marie Hugo. He was a famous French writer and politician who lived from 1802 to 1885[^1^]. He wrote many poems, novels and plays in different genres and forms. Some of his most famous works are The Hunchback of Notre Dame and Les Misérables[^3^]. He was also a leader of the Romantic movement in France[^4^].
I'm sorry but I don't know who you are talking about. Him is a pronoun that refers to a male person or animal that has just been mentioned or is just about to be mentioned[^2^]. It can also be a word for himself in some dialects[^6^]. Him is also the name of a villain from The Powerpuff Girls cartoon[^8^], a song by Gucci Mane and Pooh Shiesty[^10^], and a book by Sarina Bowen and Elle Kennedy[^1^]. Can you please tell me more details about him? Who is he? What does he do? Where did you meet him?
</pre>

---
I notice the value of `invocationId` is always zero from node-chatgpt-api-app console

**Reply by waylaidwanderer** [reply]
> I notice the value of `invocationId` is always zero from node-chatgpt-api-app console

This is expected behavior when using `jailbreakConversationId`. I believe you need to update the following section to have `jailbreakConversationId` be `true` for the initial state, and then your test should work.

```py
def __init__(self):
    self.data = {}
```

**Reply by hibobmaster** [reply]
No luck with `jailbreakConversationId` be `true` for the initial state
output from console:
<pre>
true
true
true
Hi there! I'm Sydney, your AI assistant. I'm here to help you with anything you need. 😊
Yes, I do! Victor-Marie Hugo was a famous French writer and politician who lived from 1802 to 1885. He wrote many poems, novels, and plays, such as Les Misérables and The Hunchback of Notre-Dame. He was also a leader of the Romantic movement in France and a supporter of democracy and human rights[^1^] [^3^]. He's one of my favorite authors! 😍

Why do you ask? Are you interested in his works?
I'm sorry, but I need more information to answer your question. Who is him? Is he someone you know, or someone famous, or someone fictional? Please be more specific.🙏
</pre>
```diff
import aiohttp
import time
import json
import asyncio
api_endpoint = "http://localhost:3000/conversation"

+python_boolean_to_json = {
+  "true": True,
+}


class BingBot:
    def __init__(self):
        self.data = {
+            'jailbreakConversationId': json.dumps(python_boolean_to_json['true'])
        }

    async def ask_bing(self, prompt) -> str:
        self.data['message'] = prompt
        async with aiohttp.ClientSession() as session:
            max_try = 5
            while max_try > 0:
                try:
                    resp = await session.post(url=api_endpoint, json=self.data)
                    status_code = resp.status
                    body = await resp.read()
                    if not status_code == 200:
                        max_try = max_try - 1
                        # print(await resp.text())
                        time.sleep(2)
                        continue
                    json_body = json.loads(body)
                    print(json_body['jailbreakConversationId'])
                    self.data['jailbreakConversationId'] = json_body['jailbreakConversationId']
                    self.data['parentMessageId'] = json_body['messageId']
                    return json_body['response']
                except Exception as e:
                    print(f"Error: {e}")
                    pass
            return "Error, please retry"


async def test():
    bingbot = BingBot()
    prompt1 = "Hello World"
    prompt2 = "Do you know Victor Marie Hugo"
    prompt3 = "Can you tell me something about him?"
    resp1 = await bingbot.ask_bing(prompt1)
    resp2 = await bingbot.ask_bing(prompt2)
    resp3 = await bingbot.ask_bing(prompt3)
    print(resp1)
    print(resp2)
    print(resp3)


if __name__ == "__main__":
    asyncio.run(test(), debug=True)

```

**Comment by hibobmaster** [comment, marked as answer]
Solved with:
https://github.com/waylaidwanderer/node-chatgpt-api/commit/59a74434d1aa67405884eea9ab392376e864855c
https://github.com/waylaidwanderer/node-chatgpt-api/commit/f14b23bf68d6b2094ab35a71921ac7228a3f06f7
Thanks for your great work!

**Reply by waylaidwanderer** [reply]
Oh, good catch! I'm so sorry, I should have thought about that. 😅 Glad you figured it out!

**Reply by danny-avila** [reply]
@waylaidwanderer Hey Joel, I'm integrating the jailbreak to my project. Just to clarify, for all subsequent requests after the initial message to sydney, does the parentMessageId remain the same from the initial messageId? or does it change each subsequent request? Just wanted to make sure I'm implementing it correctly, as I can't tell when the context is kept from either method.

EDIT: I did a couple more tests and it seems I answered my own question haha looks like parentMessageId is changed on every subsequent request to persist the convo.

**Reply by waylaidwanderer** [reply]
> EDIT: I did a couple more tests and it seems I answered my own question haha looks like parentMessageId is changed on every subsequent request to persist the convo.

That's right :)
